### PR TITLE
Fix #16479

### LINKED ISSUE #16479: Rails.env not set correctly in initializers when using rails console
I observed that `Rails.env` is not set correctly within initializers when using `rails console -e`.  For example:

    $ rails -v
    Rails 4.1.4
    $ rails new example
    $ cd example
    $ cat > config/initializers/example.rb <<"DOC"
    File.open("example.env", "w") {|io| io.puts Rails.env }
    DOC
    $ printf "puts Rails.env; exit" | rails console -e test
    Loading test environment (Rails 4.1.4)
    Switch to inspect mode.
    puts Rails.env; exit
    test
    $ cat example.env
    development

You can see that the initializer is picking up the 'development' as a default but the console eventually sets `Rails.env` to test before the prompt appears.  You can work around this issue by using an ENV variable instead of the option.

    $ rm example.env
    $ printf "puts Rails.env; exit" | RAILS_ENV=test rails console
    Loading test environment (Rails 4.1.4)
    Switch to inspect mode.
    puts Rails.env; exit
    test
    $ cat example.env
    test

### PR DESCRIPTION
**Pull request by skull-squadron**
Set `Rails.env` much earlier when loading `rails console` with a command-line specified environment so that initializers run in the specified environment rather than `RAILS_ENV` or `development`.

### CODE REVIEW

**Comment by steakknife** [comment]
Tested, works.

**Comment by steakknife** [comment]
This PR allows `rails runner`, `rails console` and `rails server` to consistently respect `-e {{environment}}` earlier so initializers get the right `Rails.env`.

File: railties/lib/rails/env.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    #   Rails.env.production? # => false
+    def env
+      @_env ||= ActiveSupport::StringInquirer.new(ENV["RAILS_ENV"] || ENV["RACK_ENV"] || "development")
+    end
+
+    # Sets the Rails environment.
+    #
+    #   Rails.env = "staging" # => "staging"
+    def env=(environment)
+      @_env = ActiveSupport::StringInquirer.new(environment)
```
Comment (egilburg) [inline railties/lib/rails/env.rb]: Perhaps move updating `RAILS_ENV` here as well, so rest of code can just do `Rails.env = new_env` without also assigning it to RAILS_ENV everywhere.

Comment (steakknife) [inline railties/lib/rails/env.rb]: Sounds good.

—

**Comment by steakknife** [comment]
Getter reverted to original; setter invalidates the auto-memoized variable and encapsulates the impure side-effect of setting `RAILS_ENV` as needed.

File: railties/lib/rails/commands/commands_tasks.rb [hunk 2]
```diff
@@ -59,7 +61,7 @@ def console
       options = Rails::Console.parse_arguments(argv)
 
       # RAILS_ENV needs to be set before config/application is required
-      ENV['RAILS_ENV'] = options[:environment] if options[:environment]
+      Rails.env = ENV['RAILS_ENV'] = options[:environment] if options[:environment]
```
Comment (egilburg) [inline railties/lib/rails/commands/commands_tasks.rb]: Don't need ` = ENV['RAILS_ENV']` here now

File: railties/lib/rails/commands/dbconsole.rb [hunk 3]
```diff
@@ -16,7 +17,7 @@ def initialize(arguments = ARGV)
 
     def start
       options = parse_arguments(arguments)
-      ENV['RAILS_ENV'] = options[:environment] || environment
+      Rails.env = ENV['RAILS_ENV'] = options[:environment] || environment
```
Comment (egilburg) [inline railties/lib/rails/commands/dbconsole.rb]: Don't need ` = ENV['RAILS_ENV']` here now

**Comment by steakknife** [comment]
Kinda edge-case design decision [here](https://github.com/rails/rails/pull/19255/files#diff-6ee7d5f76df3e71b81ae2bb578418215R19)

a. `ENV['RAILS_ENV'] = ...` would standardize the environment in sub-processes/-threads, but it would behave differently than currently.
b. `ENV['RAILS_ENV'] ||= ...` would retain existing behavior, but would *not* dictate the same environment for sub-processes/-threads (might be a Good Thing :tm:).

Thinking about it, I'm leaning toward b. for now to avoid breakage.

**Comment by rafaelfranca** [comment]
This seems to be a spring issue not Rails issue. See https://github.com/rails/rails/issues/16479#issuecomment-83144657

**Comment by steakknife** [comment]
The issue of unset environment appeared to be present in 4.2.0 without springified executables `rails .... -e {{environment}}` as was evidenced by `puts Rails.env` in an initializer, but repeating clean room testing revealed that the room wasn't clean enough as env is set in initializers with `-e`.





This patch is unnecessary, but the refactoring might not hurt.



—

**Pull request closed without merging**